Mood Bot

NiKnight94 opened the pull request:
Framework for the mood checking function of the mindfulness bot.
Still need to incorporate Mongo as a method of data storage. Using shelve files to store data, currently.

NiKnight94 committed:
Mood Bot - WIP

Framework for the mood checking function of the mindfulness bot.
Still need to incorporate Mongo as a method of data storage. Using shelve files to store data, currently.

NiKnight94 committed:
Updated Mood Bot

Made changes as discussed, including:
- using doc strings for comments
- lines 58-71: use a dictionary for checking the day, mapping day of the week to an integer.
- line 36: make day regex global
- combine for loops in add/del functions
- use longer variable names consistently
 --> switch d to diary
 --> switch t to time_input and time_inputs
- used fixed strings or functions for repeated messages
- defined custom exceptions (will add in another PR).

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+
+day_dictionary = {'monday': 0, 'mon': 0, 'tuesday': 1, 'tues': 1, 'tue': 1, 'wednesday': 2, 'wednes': 2, 'wed': 2,
+                  'thursday': 3, 'thurs': 3, 'thur': 3, 'friday': 4, 'fri': 4, 'saturday': 5, 'satur': 5, 'sat': 5,
+                  'sunday': 6, 'sun': 6}
```
This whole setup is great I like it

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+                'lonely', 'prideful', 'confused', 'disgusted', 'surprised', 'disappointed')
+
+"""Power is for switching the auto_checker ON/OFF"""
+auto_checker_power_switch = 'OFF'
```
This should definitely be a boolean

Quasimonomial replied:
I want to be able to 

`if auto_checker_power_switch:`

over 

`if auto_checker_power_switch == 'Foo'`

NiKnight94 replied:
Done. Switched auto_checker_power_switch to a boolean, but also turned it into a function that stores this boolean to a user-specific file so that users don't get intermingled and so that that the autochecker's status is remembered even when the bot is off.

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+    """This function waits until the next mood check is scheduled and then activates the mood check function."""
+    global auto_checker_power_switch
+    author = context.author
+    if power.upper() == 'OFF':
```
If we move `auto_checker_power_switch` to be a boolean, we can treat this with

`auto_checker_power_switch = true if power.upper() == 'ON' else false`

NiKnight94 replied:
Switched to a boolean and now the "power_switch" is a function that pulls from an external file.
' on_or_off = on_or_off.upper()
    if on_or_off == 'OFF':
        target_time = 0
        set_autochecker(author, False)
    elif on_or_off == 'ON':
        target_time = mood_time_retriever(author)
        set_autochecker(author, True)'

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+date_gate = 'Open'
+type_gate = 'Closed'
+rating_gate = 'Closed'
+why_gate = 'Closed'
```
These should probably be booleans too for a similar reason

NiKnight94 replied:
Switched to an integer like you mention in the other comment, and have stored the data to a file so not to confuse users.

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+why_gate = 'Closed'
+stage_array = ['date', 'type', 'rating', 'explanation']
+
+"""Repeated Strings"""
```
I like this a lot, I like the idea of having all of our copy in one place so we can edit text without diving deep into the code

Quasimonomial replied:
One day these should be constants `IN_ALL_CAPS`

Quasimonomial replied:
e.g. --  `DAY_FORMAT_ERROR_MESSAGE = 'Please enter the thing without being wrong'`

NiKnight94 replied:
I have capitalized the string constants and moved them to another file.

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+        day_mo = day_regex.search(day_input)
+        if day_mo is None:
+            await day_format_error(context, day_input)
+        else:
```
we could try `lower_day_input = day_input.lower()` to DRY this up a bit

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+    """ This function allows the user to add a day of the week to their mood check schedule. """
+    author = context.author
+    [mood_days, _] = mood_schedule(author)
+    for day_input in day_inputs:
```
we can `.map()` this to `.lower()`

NiKnight94 replied:
I googled and learned that you can do this in Python using the lambda function.
'day_inputs_lower = list(map(lambda x: x.lower(), day_inputs))'

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+@mood_bot.command(name='delete_day', help='Delete day(s) from mood check schedule.',
+                  aliases=['del_day', 'delday', 'deleteday', '-day', 'del_days', 'delete_days',
+                           'deldays', 'deletedays', '-days'])
+async def delete_day(context, *day_inputs):
```
We might want the day schedule to be a hash / dictionary instead of a list / array

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+        else:
+            if day_input.lower() == 'everyday':
+                for j in range(7):
+                    mood_days[j] = 0
```
with a dictionary we could do `mood_days["sun"] = false`

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+    await new_schedule_display_message(context, 'days', num_days_to_word_days(mood_days))
+
+
+async def time_not_there_error(context, mood_times, time_input):
```
lets move all these in this file to the top so they are in one place

NiKnight94 replied:
Done. Moved all the error message functions to the top of the file.

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+async def mood_status(context):
+    await context.send(f"The auto mood checker is {auto_checker_power_switch}.")
+    mood_check_status = 'has not been started.'
+    if date_gate == 'Closed':
```
this gate variable should be a integer

Quasimonomial replied:
if we have  `stage_array = ['date', 'type', 'rating', 'explanation']`

the user can have a `gate_stage` which is an integer

then, if your gate_stage is at '2', you are doing a rating

NiKnight94 replied:
I have switched the stage variable to an integer, which increments as you stated in the next comment.

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+                           "Please select from the following mood types:\n" +
+                           str(mood_choices) + "\n" +
+                           "Please respond by typing 'mood_type' followed by your selected mood.")
+        date_gate = 'Closed'
```
you can do gate +=1, and wrap it around somehow

Oh you can do
`gate = (gate + 1) % 4`

NiKnight94 replied:
I have implemented the incrementing integer with the wraparound.

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+type_gate = 'Closed'
+rating_gate = 'Closed'
+why_gate = 'Closed'
+stage_array = ['date', 'type', 'rating', 'explanation']
```
I'm a bit concerned that with this setup, we use one gate over all of our users, or worse, we do not persist the gate between commands.  We could do either storing this in mongo or more reasonably with a smaller number of users we could use a dictionary with user ids b/c this might make more sense to cache then to move to permanent storage

Quasimonomial replied:
We might want to have @Angel1412kaitou think about what happens if a user aborts the mood check in halfway through the process

NiKnight94 replied:
Currently, saving to an external file, not sure how to cache. As for the aborting process, I do have a reset function that goes off automatically if a new auto-check is started before another is finished. This function deletes the data from the incomplete mood check.

Angel1412kaitou replied:
I like Nick's idea. Sorry I've been MIA (and plan on being MIA until this semester ends) but you guys are doing great

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+            await mood_type_error(context, mood_type.lower())
+        else:
+            mood_diary_storage(author, 'type', mood_type)
+            await context.send(f"On a scale from 1 to 10, how {mood_type} do you feel?\n" +
```
we can move all these strings to constants above in all these messages, would be nice to take them out of this part of the app

NiKnight94 replied:
I have converted all strings to constants and have stored them in another file.

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+                await mood_why_message(context, mood_stamp, rating_stamp)
+
+
+@mood_bot.command(name='why', aliases=['explanation', 'explan', 'explain'])
```
maybe add a `"journal"` or `"diary"` alias

Quasimonomial commented on mood_bot.py:
```diff
@@ -0,0 +1,726 @@
+            for j in range(start, end):
+                entry = [str(date_data[j]), str(type_data[j]), str(rating_data[j]), str(why_data[j])]
+                why_string = entry[3]
+                why_string = str(why_string)
```
look up escape characters maybe? maybe that works? @PhilippRisius ?

NiKnight94 replied:
So, I was trying to '.join' the 'why' statement in this function which led to all these extra punctuation for some reason. But, when I use the '.join' function in the why function, it works properly and converts the list to a string.

NiKnight94 replied:
So, I no longer have all these replace functions.

NiKnight94 committed:
Added changed as prescribed by Travis

- Change auto checker switch to a boolean<
    probably switch to auto_checker_on<
- Change days of the week from 1 and 0 to booleans<
- Change repeated strings to constants in ALL CAPS<
- Factor out the lower using map for day input<
    - didn't use map, used a for loop<
- Reassigned power and mood variables as their upper/lower versions<
- Move error functions into one spot<
- Change gates to an incrementing integer<
- Change stage_array to a tuple<
- How does the bot know whose stage of the moodcheck is whose?<
- Move all string messages to one spot for export to another file<
    - used %s string interpolation to handle variables<
- Added a few additional aliases and helps<
- Figure out alternative to replace<
    - used string join in the mood_why function (before storage)<
- Ran into a problem:
    - cannot call the auto mood check function without calling the function
    - tried making it a separate event, but to no avail
